### PR TITLE
Added support to access native posix flash via host file system

### PR DESCRIPTION
**Pull request by vanwinkeljan**
Add support to access native posix flash device via host file system.
The flash device is mapped into the host file system by means of the FUSE library.

These changesets where original part of PR #12503

edit: actually only access to FS partitions is provided

### CODE REVIEW

File: soc/posix/inf_clock/Kconfig.soc
Comment (aescolar) [inline soc/posix/inf_clock/Kconfig.soc]: This dependency would not be correct, or? Doesn't it depend just on ARCH_POSIX? (or maybe also in FLASH_NATIVE_POSIX)

**Review by aescolar — DISMISSED** [review]
Please move the code, kconfig and cmake into drivers/flash
The Kconfig can be into drivers/flash/Kconfig.native_posix
The cmake chunk into drivers/flash/CMakeLists.txt
The code could be just named drivers/flash/fuse_native_posix.c or so

**Comment by vanwinkeljan** [comment]
> Please move the code, kconfig and cmake into drivers/flash

I did try this but it is not an option as the compiler flags need to be changed to make use of FUSE and there is a conflict with the flags used in the zephyr flash library.

File: soc/posix/inf_clock/Kconfig.soc
Comment (vanwinkeljan) [reply soc/posix/inf_clock/Kconfig.soc]: As it is part of inf_clock it did look more logic to make it depend on SOC_POSIX.
Further the FUSE support has no direct dependency on the flash device as it uses the file system API (``depends on FILE_SYSTEM``) to access the  flash device.

Comment (aescolar) [reply soc/posix/inf_clock/Kconfig.soc]: But, the dependency would still be only on ARCH_POSIX? (not SOC_POSIX)

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]: How about:
```
  The flash content can be accessed from the host system, as explained
  in the `Host based flash access`_ section.
```

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]:
```suggestion
If a flash device is present, the file system partitions on the flash
```

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]:
```suggestion
the required UNIX file system calls, and provides access to the flash file
```

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]:
```suggestion
system partitions with normal operating system commands such as
``cd``, ``ls``, and ``mkdir``.
```

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]:
```suggestion
option *--flash-mount*. As this directory operates as a mount point for FUSE,
```

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]:
```suggestion
On exit, the native POSIX board application will take care of unmounting the
```

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]:
```suggestion
crashes, you can cleanup the stale mount point by using the program ``fusermount``::
```
and delete the ``eg::`` on the next line.

**Review by dbkinder — CHANGES_REQUESTED** [review]
Thanks for adding the documentation.  I've got a few recommended edits.

**Comment by aescolar** [comment]
> there is a conflict with the flags used in the zephyr flash library.

@vanwinkeljan Would declaring the fuse part as a separate library do the trick? (even if it is in the same folder as the other flash drivers)

File: samples/subsys/shell/fs/prj.conf
Comment (jarz-nordic) [inline samples/subsys/shell/fs/prj.conf]: Why you need `CONFIG_CONSOLE=y` ?
It is not needed anymore to use new shell.

Comment (vanwinkeljan) [reply samples/subsys/shell/fs/prj.conf]: There is no need, will be removed.

**Comment by vanwinkeljan** [comment]
> Would declaring the fuse part as a separate library do the trick? (even if it is in the same folder as the other flash drivers)

Yes it does, moved the code to subsys/fs

File: soc/posix/inf_clock/Kconfig.soc
Comment (vanwinkeljan) [reply soc/posix/inf_clock/Kconfig.soc]: changed dep to ARCH_POSIX (in subsys/fs/KConfig)

**Comment by vanwinkeljan** [comment]
Renamed fuse_flash to fuse_fs_access as the code actually only provides access to the FS partitions

**Comment by aescolar** [comment]
@vanwinkeljan I guess you have tested that it works nicely?
You may want to add a line in CODEOWNERS for this new fuse file assigned to yourself.
If you commit to maintain/support this new functionality I think it should be ok to merge it, as there is very little risk, this being a new optional feature.
Regarding the new fs function, I hope @nvlsianpu has taken a look(?) and is ok with it

File: subsys/fs/fuse_fs_access.c
Comment (aescolar) [inline subsys/fs/fuse_fs_access.c]: nit: you may want `posix_print_trace()` to be sure this ends up in the same place as all other prints in the posix arch

File: boards/posix/native_posix/doc/index.rst
Comment (aescolar) [inline boards/posix/native_posix/doc/index.rst]: ``doc/_build/rst/doc/boards/posix/native_posix/doc/index.rst:642: WARNING: Unexpected indentation.``

Comment (aescolar) [reply boards/posix/native_posix/doc/index.rst]: @vanwinkeljan This is how it looks: https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/16467/docs/boards/posix/native_posix/doc/index.html#host-based-flash-access

File: samples/subsys/shell/fs/README.rst
Comment (aescolar) [inline samples/subsys/shell/fs/README.rst]: fs mount nf-s-+f+s /nffs

**Comment by aescolar** [comment]
@vanwinkeljan Apart from those minor issues I'm ok with this, and can approve it after you fixed them

File: boards/posix/native_posix/doc/index.rst
Comment (dbkinder) [inline boards/posix/native_posix/doc/index.rst]: There's an extra space at the front of this line (642) that you should delete :)

Comment (dbkinder) [reply boards/posix/native_posix/doc/index.rst]: P.S.  using ``make htmldocs-fast`` avoids creating all the options documentation details and runs in a quarter of the time as the full doc build.

**Review by dbkinder — CHANGES_REQUESTED** [review]
one fix to remove the doc-build warning...

File: samples/subsys/shell/fs/README.rst
Comment (vanwinkeljan) [reply samples/subsys/shell/fs/README.rst]: nice catch

File: boards/posix/native_posix/doc/index.rst
Comment (vanwinkeljan) [reply boards/posix/native_posix/doc/index.rst]: @dbkinder is this a new make target, as I get `No rule to make target`?

**Comment by vanwinkeljan** [comment]
> You may want to add a line in CODEOWNERS for this new fuse file assigned to yourself.

Done

> If you commit to maintain/support this new functionality I think it should be ok to merge it, 

No problem with that, I think it is an added value for zephyr

**Comment by dbkinder** [comment]
The Makefile in the Zephyr root has this htmldocs-fast target.

**Review by aescolar — APPROVED** [review]

**Comment by nvlsianpu** [comment]
Any chance that this will work on widowns?

File: include/fs.h
Comment (nvlsianpu) [inline include/fs.h]: what is end-of-mount condition?

File: subsys/fs/fuse_fs_access.c
Comment (nvlsianpu) [inline subsys/fs/fuse_fs_access.c]: errcode ignored, cast to void or serve return values

File: subsys/fs/fuse_fs_access.c
Comment (nvlsianpu) [inline subsys/fs/fuse_fs_access.c]: I think that is what vfs requires, not NFFS itself. I see vfs calls bellow, not direct NFFS calls.

File: subsys/fs/fuse_fs_access.c
Comment (nvlsianpu) [inline subsys/fs/fuse_fs_access.c]: Doses NFFS supports truncate at all? I thought it doesn't.

**Comment by vanwinkeljan** [comment]
@nvlsianpu
> Any chance that this will work on widowns?

No currently this not possible due to 2 limitations:
* Fuse is not suported on windows
* The WSL layer does not support 32-bit application, and native_posix is a 32-bit appilcation

Now it looks that both issues will be solved by introduction of WSL2, an initial build of WSL2 would be available by end of June for people with access to WIndows Insider.
https://devblogs.microsoft.com/commandline/announcing-wsl-2/

File: include/fs.h
Comment (vanwinkeljan) [reply include/fs.h]: Correct this would be more clear, I will update the code. 

I will also add a comment to indicate that the parameter `number` will be incremented if a valid mount point is returned.

File: subsys/fs/fuse_fs_access.c
Comment (vanwinkeljan) [reply subsys/fs/fuse_fs_access.c]: correct this required by vfs and not nffs, will update comment

File: subsys/fs/fuse_fs_access.c
Comment (vanwinkeljan) [reply subsys/fs/fuse_fs_access.c]: NFFS does not support truncate but correctly returns -ENOTSUP so FUSE will return this correctly to the host user space program. Note that the FAT file system implementation in zephyr does support truncate.

File: include/fs.h
Comment (vanwinkeljan) [reply include/fs.h]: ENOEXIST does not exist in zephyrs errno.h file using ENOENT instead.

**Review by nvlsianpu — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
docs LGTM, thanks!

**Pull request merged**
